### ISSUE TITLE
musl libc & Alpine based Docker images compatibility

### ISSUE THREAD

**Issue by ghost** [deeplearning4j/deeplearning4j#9715]
#### DeepLearning4J is incompatible with musl libc & Alpine based Docker images

I tried running a DL4J application inside a Docker container that is based on an Alpine Linux Image (eclipse-temurin:8-alpine). When starting the application, I get a fatal error. The Java VM dies.

**expected behavior:** I can run a DL4J application inside a musl libc based Linux Docker container like Alpine
**encountered behavior:** When I run my DL4J application in a Docker container that uses Alpine Linux as its base image, I get a fatal error.

```
14:45:20.692 [main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000000000021c6, pid=8, tid=0x00007f3c87a8cb38
#
# JRE version: OpenJDK Runtime Environment (8.0_332-b09) (build 1.8.0_332-b09)
# Java VM: OpenJDK 64-Bit Server VM (25.332-b09 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  0x00000000000021c6
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /opt/dl4j/hs_err_pid8.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
#
Aborted
```
I attached the file [hs_err_pid8.log](https://github.com/eclipse/deeplearning4j/files/8949956/hs_err_pid8.log)

**One suspicious log entry is:**

```
R12=0x00007f3c65aaa558: <offset 0x23e558> in /root/.javacpp/cache/example-1.0.0-SNAPSHOT-shaded.jar/org/bytedeco/openblas/linux-x86_64/libquadmath.so.0 at 0x00007f3c6586c000
R13=0x00007f3c8891a880: <offset 0x96880> in /lib/ld-musl-x86_64.so.1 at 0x00007f3c88884000
R14=0x00007f3c87a8cb38 is an unknown value
```

I guess `/root/.javacpp/cache/example-1.0.0-SNAPSHOT-shaded.jar/org/bytedeco/openblas/linux-x86_64/libquadmath.so.0` is linked against glibc, but on Alpine Linux there is no glibc. There is gcompat, but that doesn't fix this issue.

I ran:

```
objdump -x /root/.javacpp/cache/example-1.0.0-SNAPSHOT-shaded.jar/org/bytedeco/openblas/linux-x86_64/libquadmath.so.0
```

and it shows:

```
Version References:
  required from libm.so.6:
    0x09691a75 0x00 05 GLIBC_2.2.5
  required from libc.so.6:
    0x0d696914 0x00 08 GLIBC_2.4
    0x0d696913 0x00 07 GLIBC_2.3
    0x09691a75 0x00 06 GLIBC_2.2.5
    0x06969190 0x00 04 GLIBC_2.10
```
So I guess these libraries were actually compiled and linked with glibc.

#### Version Information

* Deeplearning4j version: 1.0.0-M2
* Platform information (OS, etc): Alpine Linux v3.15.4, libc:musl - unknown musl - unknown

#### Additional Information

I attached an example project ([example.zip](https://github.com/eclipse/deeplearning4j/files/8950960/example.zip)) and corresponding Dockerfiles to replicate the problem. There is a README.md inside with instructions.

Thanks in advance.

Regards,
Andreas

**Comment by agibsonccc** [comment]
@schipplock you'd have to build at least a few things that aren't based on glibc I believe. We heavily depend on a lot of javacpp projects: https://github.com/bytedeco/javacpp-presets - I don't really ever see people ask for musl.
There's also the c++ lib itself which uses glibc everywhere. 

That's fixable building from source but your mileage will vary with the different depdendeices. What subset of this are you hoping to get working? 

Just running a model on a cpu?

**Comment by agibsonccc** [comment]
@schipplock just following up but sorry I seriously can't make this a priority without more information and clear use cases. I know a number of folks use musl but it would basically require a separate classifier for our c++ library at a minimum.
Our builds are under .github/workflows if you want to take a look.

What I might be able to do is publish a separate classifier, but even then I need to find a way to make openblas work. That won't solve this problem for all backends either. 

Our openblas that we depend on upstream by default uses glibc already. The best I can recommend if you want to help here would be to help us publish a separate musl that builds also openblas linked against musl. I'm not sure how it would work but you would also have to submit a separate issue upstream at: https://github.com/bytedeco/javacpp-presets for an openblas presets. 

Chances are due to the marketshare though, you'll probably just get told to send a PR in a similar fashion to what I'd be asking you to do here. I'd be happy to guide you on our end though if you want to tackle this.

**Comment by ghost** [comment]
Hi @agibsonccc,
thanks for your answer.

I can see this is a lot of work, and I'm not asking nor expecting that anyone is  making it work on musl based environments :). I have to admit: I was expecting someone telling me that musl is simply not supported. I'm not the biggest musl fan either, simply because of compatibility reasons. This here is a good example. People build, link and test their stuff with glibc.

In the meantime, I could convince Payara to provide glibc based Docker images in the future :). My employer is paying for Payara Enterprise, it's not paying anything for DL4J.

Again, thanks for your feedback.

Regards,
Andreas

**Comment by agibsonccc** [comment]
@schipplock depending on your team's needs commercial engagements are possible. Please ping support@konduit.ai if you think getting a dedicated build + some other support might be interesting. Thanks for understanding!  Helping the jarkata ee ecosystem is definitely of interest to us as a fellow eclipse project but I also understand the trade offs your team has to make.

**Comment by ghost** [comment]
Hi Adam,
we are more interested in "some other support". I'll forward this to my collegue (I'm just the poor IT guy :P).

Regards,
Andreas
